test/functional/p2p_filter.py

```diff
@@ -104,6 +106,21 @@ def run_test(self):
             txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 7)
             filter_node.wait_for_tx(txid)
 
+        self.log.info('Check that request for filtered blocks is ignored if no filter is set')
+        filter_node.merkleblock_received = False
+        filter_node.tx_received = False
+        with self.nodes[0].assert_debug_log(expected_msgs=['received getdata']):
+            block_hash = self.nodes[0].generatetoaddress(1, self.nodes[0].getnewaddress())[0]
+            filter_node.wait_for_inv([CInv(MSG_FILTERED_BLOCK, int(block_hash, 16))])
```

Comment:
Sorry, I found a bug in my test code.

```diff
diff --git a/test/functional/p2p_filter.py b/test/functional/p2p_filter.py
index 39050d98dd..7f8da2ba90 100755
--- a/test/functional/p2p_filter.py
+++ b/test/functional/p2p_filter.py
@@ -37,8 +37,9 @@ class FilterNode(P2PInterface):
         for i in message.inv:
             # inv messages can only contain TX or BLOCK, so translate BLOCK to FILTERED_BLOCK
             if i.type == MSG_BLOCK:
-                i.type = MSG_FILTERED_BLOCK
-            want.inv.append(i)
+                want.inv.append(CInv(MSG_FILTERED_BLOCK, i.hash))
+            else:
+                want.inv.append(i)
         if len(want.inv):
             self.send_message(want)
 
@@ -105,7 +106,7 @@ class FilterTest(BitcoinTestFramework):
         filter_node.tx_received = False
         with self.nodes[0].assert_debug_log(expected_msgs=['received getdata']):
             block_hash = self.nodes[0].generatetoaddress(1, self.nodes[0].getnewaddress())[0]
-            filter_node.wait_for_inv([CInv(MSG_FILTERED_BLOCK, int(block_hash, 16))])
+            filter_node.wait_for_inv([CInv(MSG_BLOCK, int(block_hash, 16))])
             filter_node.sync_with_ping()
             assert not filter_node.merkleblock_received
             assert not filter_node.tx_received

Reply:
Reason is that invs are only allowed to be BLOCK or TX, nothing else

Reply:
Oh, that's true. But I'm a bit confused on why it was still possible that the test passed? Was the code in `on_inv()` directly modifying the object that was stored in the `last_message` dictionary?

Reply:
Yes, in python everything is a reference (like a shared pointer, but as a reference). So when modifying a member variable or an element in a list or so, it is going to modify all other instances that point to that list or object.

Reply:
The test passed because it checked for `MSG_FILTERED_BLOCK`, whereas it should have tested for `MSG_BLOCK`

Reply:
Thanks for the explanation, that makes sense. I didn't consider that in Python every name is a reference :see_no_evil: